### PR TITLE
Extract phrases to put in feedback explorer tool

### PR DESCRIPTION
**Pull request by ff-l**
Building off previous iteration, extract verb-based phrases from feedback (1-10/04) and put in columns. 
## Main difficulty
Identify exact match of extracted phrase within original comment text. 
Context: we're extracting phrases from text, where phrases are a few concatenated words. As we're not trying to match every single part of speech there exists, some information is lost. 

- From "I wanted to quickly check out your employment advice" you get stuff like "wanted to check employment advice".
- Annoyingly, due to tokenization, stuff like "can't" is also separately split to ca n't.

To sort this out I ended up going with producing ngrams of original comment (n being a bit larger than the token length of the phrase we're trying to match). Once finding the maximal and optimal match of ngram to phrase we identify the submatch we want, by matching first and last token of phrase. 
Then from ngram "try to get your recommendations" we can match phrase "try to get recommendations" on "(try).*(get)"

### CODE REVIEW

File: notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb [hunk 1]
```diff
@@ -0,0 +1,1339 @@
...
+   "outputs": [],
+   "source": [
+    "# !pip3 install wordcloud\n",
+    "# !pip3 install polyglot\n",
+    "# !pip3 install pyicu\n",
+    "# !pip3 install pycld2\n",
+    "# !pip3 install morfessor\n",
+    "# !pip3 install polyglot\n",
+    "# !pip3 install fuzzywuzzy\n",
+    "# !pip3 install seaborn"
```
Comment (ESKYoung) [inline notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb:16]: Consider updating the `requirements-dev.txt` file with the packages used in this Jupyter notebook.

File: notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb [hunk 2]
```diff
@@ -0,0 +1,1339 @@
...
+   "source": [
+    "stop_words = list(stopwords.words('english'))\n",
+    "punctuation = list(string.punctuation) + ['’']\n",
+    "token_blacklist = stop_words + punctuation + pii_filtered\n",
+    "\n",
+    "def split_sentences(comment):\n",
+    "    return nltk.sent_tokenize(comment)\n",
+    "\n",
+    "def remove_stopwords_punctation(sentences):\n",
+    "    return [[(t[0], t[1], t[2]) for t in sent if t[0].lower() not in token_blacklist] for sent in sentences]\n",
```
[the remark this answers is not in the document]
Comment (ff-l) [reply notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb:146]: Remainder from the past, will remove

File: notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb [hunk 3]
```diff
@@ -0,0 +1,1339 @@
...
+   "source": [
+    "df['pos_tag'] = df[['Q3_pii_removed', 'is_en']].progress_apply(lambda x: part_of_speech_tag(x[0]) \n",
+    "                                                     if x[1] else [], axis=1)\n",
+    "df['lemmas'] = df['pos_tag'].progress_map(lambda x: [token[2] for sent in x for token in sent])\n",
+    "\n",
+    "df['words'] = df['pos_tag'].progress_map(lambda x: [token[0] for sent in x for token in sent])\n",
+    "\n",
+    "df.to_csv(os.path.join(DATA_DIR, \"uis_20200401_20200409_lang_pos.csv\"), index=False)\n",
+    "df = pd.read_csv(os.path.join(DATA_DIR, \"uis_20200401_20200409_lang_pos.csv\"))\n",
+    "df['pos_tag'] = df['pos_tag'].map(literal_eval)"
```
[the remark this answers is not in the document]
Comment (ff-l) [reply notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb:256]: Thank you, I was being lazy

File: notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb [hunk 4]
```diff
@@ -0,0 +1,1339 @@
...
+    "regex_for_theme(\"stats\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def regex_group_verbs(verb):\n",
```
[the remark this answers is not in the document]
Comment (ff-l) [reply notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb:601]: Agreed, how about one for the next iteration?

File: notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb [hunk 5]
```diff
@@ -0,0 +1,1339 @@
...
+    "\n",
+    "def remove_stopwords_punctation(sentences):\n",
+    "    return [[(t[0], t[1], t[2]) for t in sent if t[0].lower() not in token_blacklist] for sent in sentences]\n",
+    "\n",
+    "def replace_pii_regex(text):\n",
+    "    return re.sub(pii_regex, \"\", text)\n",
+    "\n",
+    "def part_of_speech_tag(comment):\n",
+    "    sentences = split_sentences(comment)\n",
+    "    return [[(token.text, token.tag_, token.lemma_) for token in nlp(sentence)] for sentence in sentences]"
```
[the remark this answers is not in the document]
Comment (ff-l) [reply notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb:153]: This is by design, need to retain sentence boundaries (inner list)

File: notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb [hunk 6]
```diff
@@ -0,0 +1,1339 @@
...
+   "outputs": [],
+   "source": [
+    "df['user_phrases_dict'] = df[['theme_mentions_user', \"Q3_x_edit\"]][:].\\\n",
+    "            progress_apply(lambda x: [find_needle(phrase, x[1].lower()) for phrase  in x[0]], axis=1)\n",
+    "df['user_phrases_list'] = df['user_phrases_dict'].progress_map(lambda x: [value for phrase_dict in x \n",
+    "                                                                 for value in phrase_dict.values() \n",
+    "                                                                 if value is not None]\n",
+    "                                                if not isinstance(x, float) else [])\n",
+    "\n",
+    "df['user_phrases'] = df['user_phrases_list'].progress_map(lambda x: \", \".join(x))"
```
[the remark this answers is not in the document]
Comment (ff-l) [reply notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb:1240]: Definitely needs to be more robust + exclude non-people related nouns

File: notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb [hunk 7]
```diff
@@ -0,0 +1,1339 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "user_groups = Counter()\n",
+    "for vals in df[df['theme_mentions_user'].str.len()>0].theme_mentions_user.values:\n",
+    "    for val in vals:\n",
+    "        user_groups[val] +=1\n",
+    "user_groups.most_common(10), \"housebound\" in user_groups, len(user_groups)"
```
[the remark this answers is not in the document]
Comment (ff-l) [reply notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb:1101]: I'm getting a true in my case 🤔

File: requirements-dev.txt [hunk 8]
```diff
@@ -1,2 +1,148 @@
...
+polyglot==16.7.4
+preshed==3.0.2
+prometheus-client==0.7.1
+prompt-toolkit==3.0.5
+protobuf==3.11.3
+ptyprocess==0.6.0
+py==1.8.1
+pycld2==0.41
+Pygments==2.6.1
+PyICU==2.4.3
```
[the remark this answers is not in the document]
Comment (ff-l) [reply requirements-dev.txt:93]: Updated the readme :)

File: notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb [hunk 9]
```diff
@@ -0,0 +1,1339 @@
...
+    "        \n",
+    "        print(\"verbs:\\n\", verbs[0:10])\n",
+    "        print(\"arguments:\\n\", arguments[0:10])\n",
+    "        \n",
+    "        row_list.append({\"Generic phrase\": f\"{pattern[0]} - {pattern[1]}\",\n",
+    "                         \"# of times occured\": total_occ,\n",
+    "                        \"Verb category\": f\"{pattern[0]}\",\n",
+    "                         \"Total # of verbs\" : sum(local_verbs.values()),\n",
+    "                         \"# of unique verbs\" : len(local_verbs),\n",
+    "                        \"Verb values\": \"\\n\".join(verbs[0:10]),\n",
```
[the remark this answers is not in the document]
Comment (ESKYoung) [reply notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb:925]: Discussed with @ff-l - this is used for spot checking at the moment, so leaving this for now.

File: notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb [hunk 10]
```diff
@@ -0,0 +1,1339 @@
...
+    "        print(\"====\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pd.DataFrame(row_list)[0:20].to_csv(os.path.join(DATA_DIR, \"generic_phrase_counts_top_20.csv\"), index=False)"
```
[the remark this answers is not in the document]
Comment (ESKYoung) [reply notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb:939]: See https://github.com/alphagov/govuk-corona-analysis/pull/22#discussion_r419316621

File: notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb [hunk 11]
```diff
@@ -0,0 +1,1339 @@
...
+    "        row_list.append({\"Generic phrase\": f\"{pattern[0]} - {pattern[1]}\",\n",
+    "                         \"# of times occured\": total_occ,\n",
+    "                        \"Verb category\": f\"{pattern[0]}\",\n",
+    "                         \"Total # of verbs\" : sum(local_verbs.values()),\n",
+    "                         \"# of unique verbs\" : len(local_verbs),\n",
+    "                        \"Verb values\": \"\\n\".join(verbs[0:10]),\n",
+    "                        \"Argument category\": f\"{pattern[1]}\",\n",
+    "                        \"Total # of arguments\" : sum(local_arguments.values()),\n",
+    "                         \"# of unique arguments\" : len(local_arguments),\n",
+    "                        \"Argument values\": \"\\n\".join(arguments[0:10])})\n",
```
[the remark this answers is not in the document]
Comment (ESKYoung) [reply notebooks/phrase-level-analysis/detect_extract_phrases_for_tool.ipynb:929]: See https://github.com/alphagov/govuk-corona-analysis/pull/22#discussion_r419316621

**Comment by ESKYoung** [comment]
Reviewed the code, looks good - few outstanding queries, but nothing to prevent approval. :shipit:

**Pull request merged**